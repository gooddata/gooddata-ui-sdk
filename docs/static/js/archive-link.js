document.addEventListener('DOMContentLoaded', function () {
    function createArchiveLink() {
        const link = document.createElement('a');
        link.className = 'dropdown-item';
        
        const currentHostname = window.location.hostname;
        if (currentHostname.includes('netlify.app')) {
            // For Netlify previews, link to the versions page within the current preview site
            link.href = window.location.origin + '/docs/gooddata-ui/versions/';
        } else if (currentHostname === 'localhost' || currentHostname === '127.0.0.1') {
            // For local testing, link to the versions page
            link.href = '/versions/';
        } else {
            // For production - link to the unversioned archive page
            link.href = 'https://www.gooddata.com/docs/gooddata-ui/versions/';
        }
        
        // Safe DOM manipulation - XSS protected
        const icon = document.createElement('i');
        icon.className = 'fas fa-archive me-2';
        icon.setAttribute('aria-hidden', 'true');
        
        link.appendChild(icon);
        link.appendChild(document.createTextNode('Archive'));
        
        // Add descriptive text for screen readers
        link.setAttribute('aria-label', 'View all versions in archive');
        
        return link;
    }
    
    // Select the dropdown menu and add the archive link
    const dropdownMenus = document.getElementsByClassName('dropdown-menu');
    for (let i = 0; i < dropdownMenus.length; i++) {
        const archiveLink = createArchiveLink();
        dropdownMenus[i].appendChild(archiveLink);
    }
});