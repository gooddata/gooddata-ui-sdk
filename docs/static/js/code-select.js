(function ($) {
    $(document).ready(function () {
        $(".gd-docs-code-select__tab").on("click", setActive);
        $(".gd-docs-code-select__osdetection .gd-docs-code-select__tab").each(osDetection);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");
            $(this)
                .addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-code-select__code")
                .removeClass("active")
                .filter(".lang-" + $(this).data("lang"))
                .addClass("active");
        }
    }

    var osDetection = function() {
        if(window.navigator.userAgent.indexOf($(this).data("user-agent")) >= 0) {
            $(this).trigger("click");
        }
    }

}(jQuery));