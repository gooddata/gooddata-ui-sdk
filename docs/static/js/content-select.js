(function ($) {
    $(document).ready(function () {
        $(".gd-docs-content-select__tab").on("click", setActive);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");

            $(this).addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-content__block")
                .removeClass("active")
                .filter(".content-" + $(this).data("content"))
                .addClass("active");
        }
    }
}(jQuery));
