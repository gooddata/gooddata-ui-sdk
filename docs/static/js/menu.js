$(document).ready(function () {
    var menu = $(".gd-docs-menu"),
        menuActive = $(".gd-docs-menu-page.active");

    if(menuActive.length > 0) {
        menu.scrollTop(menuActive.offset().top - menu.offset().top - parseInt(menu.css("paddingTop")));
    }

    menu.removeClass("init");
});