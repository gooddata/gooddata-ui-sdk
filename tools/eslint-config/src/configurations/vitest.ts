// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
