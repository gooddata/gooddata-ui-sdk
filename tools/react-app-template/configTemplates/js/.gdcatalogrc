{
  "hostname": "https://public-examples.gooddata.com",
  "workspaceId": "demo",
  "output": "src/catalog.ts",
  "backend": "tiger",
  "demo": true
}