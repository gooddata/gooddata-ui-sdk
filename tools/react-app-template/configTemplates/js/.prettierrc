{
    "printWidth": 110,
    "tabWidth": 4,
    "trailingComma": "all"
}
