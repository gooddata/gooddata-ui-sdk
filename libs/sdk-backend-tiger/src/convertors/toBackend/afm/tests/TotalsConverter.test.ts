// (C) 2021-2024 GoodData Corporation

import {
    bucketSetTotals,
    defSetDimensions,
    defWithDimensions,
    emptyDef,
    IExecutionDefinition,
    MeasureGroupIdentifier,
    newBucket,
    newDefForBuckets,
    newDimension,
    newTotal,
    newTwoDimensional,
} from "@gooddata/sdk-model";
import { ReferenceMd } from "@gooddata/reference-workspace";
import { describe, expect, it } from "vitest";

import { convertTotals } from "../TotalsConverter.js";

const TotalSum1 = newTotal("sum", "m1", "localAttr1");
const TotalSum2 = newTotal("sum", "m2", "localAttr1");
const TotalSum3 = newTotal("sum", "m3", "localAttr1");
const TotalMin1 = newTotal("min", "m1", "localAttr1");
const TotalMax2 = newTotal("max", "m2", "localAttr1");
const TotalNat1 = newTotal("nat", "m1", "localAttr1");
const TotalOtherDim = { ...TotalSum1, attributeIdentifier: "localAttr2" };

const Test0 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1", "localAttr2"], []),
    newDimension([MeasureGroupIdentifier]),
);
const Test1 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1", "localAttr2"], [TotalSum1]),
    newDimension([MeasureGroupIdentifier]),
);
const Test2 = defWithDimensions(
    emptyDef("test"),
    newDimension([MeasureGroupIdentifier, "localAttr2"]),
    newDimension(["localAttr1"], [TotalSum1]),
);
const Test3 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1", "localAttr2"], [TotalSum1, TotalMin1]),
    newDimension(["localAttr3", MeasureGroupIdentifier]),
);
const Test4 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1"], [TotalSum1, TotalSum2, TotalSum3]),
    newDimension(["localAttr3", MeasureGroupIdentifier]),
);
const Test5 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1"], [TotalSum1, TotalSum2, TotalMin1, TotalMax2]),
    newDimension([MeasureGroupIdentifier]),
);
const Test6 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1"], [TotalSum1]),
    newDimension([MeasureGroupIdentifier]),
    newDimension(["localAttr2"], [TotalOtherDim]),
);
const Test7 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1", "localAttr2"], [TotalSum1, TotalOtherDim]),
    newDimension(["localAttr3", MeasureGroupIdentifier]),
);
const Test8 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1", MeasureGroupIdentifier], [TotalSum1]),
    newDimension(["localAttr3"]),
);
const Test9 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1"], [TotalSum1]),
    newDimension(["localAttr3"]),
);
const Test10 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1", "localAttr2", MeasureGroupIdentifier], [TotalSum1]),
);
// Grand totals
const total = newTotal("sum", ReferenceMd.WinRate, ReferenceMd.Account.Name);
const columnTotal = newTotal("sum", ReferenceMd.WinRate, ReferenceMd.Department.Default);
const Test11 = defSetDimensions(
    newDefForBuckets("test", [
        newBucket("measure", ReferenceMd.WinRate),
        bucketSetTotals(newBucket("attribute", ReferenceMd.Account.Name), [total]),
        bucketSetTotals(newBucket("columns", ReferenceMd.Department.Default), [columnTotal]),
    ]),
    newTwoDimensional(
        [ReferenceMd.Account.Name, total],
        [ReferenceMd.Department.Default, MeasureGroupIdentifier, columnTotal],
    ),
);
// Marginal totals
const subtotal = newTotal("sum", ReferenceMd.WinRate, ReferenceMd.ForecastCategory);
const columnSubTotal = newTotal("sum", ReferenceMd.WinRate, ReferenceMd.IsActive);
const Test12 = defSetDimensions(
    newDefForBuckets("test", [
        newBucket("measure", ReferenceMd.WinRate),
        bucketSetTotals(newBucket("attribute", ReferenceMd.Account.Name, ReferenceMd.ForecastCategory), [
            subtotal,
        ]),
        bucketSetTotals(newBucket("columns", ReferenceMd.Department.Default, ReferenceMd.IsActive), [
            columnSubTotal,
        ]),
    ]),
    newTwoDimensional(
        [ReferenceMd.Account.Name, ReferenceMd.ForecastCategory, subtotal],
        [ReferenceMd.Department.Default, ReferenceMd.IsActive, MeasureGroupIdentifier, columnSubTotal],
    ),
);
// Row total + column subtotal
const Test13 = defSetDimensions(
    newDefForBuckets("test", [
        newBucket("measure", ReferenceMd.WinRate),
        bucketSetTotals(newBucket("attribute", ReferenceMd.Account.Name, ReferenceMd.ForecastCategory), [
            total,
        ]),
        bucketSetTotals(newBucket("columns", ReferenceMd.Department.Default, ReferenceMd.IsActive), [
            columnSubTotal,
        ]),
    ]),
    newTwoDimensional(
        [ReferenceMd.Account.Name, ReferenceMd.ForecastCategory, subtotal],
        [ReferenceMd.Department.Default, ReferenceMd.IsActive, MeasureGroupIdentifier, columnSubTotal],
    ),
);
// Column total + row subtotal
const Test14 = defSetDimensions(
    newDefForBuckets("test", [
        newBucket("measure", ReferenceMd.WinRate),
        bucketSetTotals(newBucket("attribute", ReferenceMd.Account.Name, ReferenceMd.ForecastCategory), [
            subtotal,
        ]),
        bucketSetTotals(newBucket("columns", ReferenceMd.Department.Default, ReferenceMd.IsActive), [
            columnTotal,
        ]),
    ]),
    newTwoDimensional(
        [ReferenceMd.Account.Name, ReferenceMd.ForecastCategory, subtotal],
        [ReferenceMd.Department.Default, ReferenceMd.IsActive, MeasureGroupIdentifier, columnSubTotal],
    ),
);

const Test15 = defWithDimensions(
    emptyDef("test"),
    newDimension(["localAttr1", "localAttr2"], [TotalNat1]),
    newDimension([MeasureGroupIdentifier]),
);

// subtotals, with unsorted totals
const Test16 = defSetDimensions(
    newDefForBuckets("test", [
        newBucket("measure", ReferenceMd.WinRate, ReferenceMd.Won),
        newBucket("attribute", ReferenceMd.Account.Name, ReferenceMd.ForecastCategory),
        bucketSetTotals(newBucket("columns", ReferenceMd.Department.Default, ReferenceMd.IsActive), [
            newTotal("sum", ReferenceMd.Won, ReferenceMd.Department.Default),
            newTotal("sum", ReferenceMd.WinRate, ReferenceMd.Department.Default),
        ]),
    ]),
    newTwoDimensional(
        [ReferenceMd.Account.Name, ReferenceMd.ForecastCategory],
        [
            ReferenceMd.Department.Default,
            ReferenceMd.IsActive,
            MeasureGroupIdentifier,
            // it should be ordered as WinRate, Won in the snapshot
            newTotal("sum", ReferenceMd.Won, ReferenceMd.Department.Default),
            newTotal("sum", ReferenceMd.WinRate, ReferenceMd.Department.Default),
        ],
    ),
);

describe("convertTotals", () => {
    const Scenarios: Array<[string, IExecutionDefinition]> = [
        ["no totals", Test0],
        ["attributes in first dim and measure group in second dim", Test1],
        ["measure group and attribute in one dimension", Test2],
        ["two totals with different functions", Test3],
        ["three totals with the same function", Test4],
        ["multiple totals with various functions", Test5],
        ["three dimensions", Test6],
        ["subtotal", Test7],
        ["total on dimension with measure group", Test8],
        ["total without any measure group", Test9],
        ["total on single dimension result spec", Test10],
        ["two totals and grand total", Test11],
        ["two subtotals and marginal total", Test12],
        ["row total and column subtotal", Test13],
        ["column total and row subtotal", Test14],
        ["subtotals with non-sorted totals", Test16],
    ];
    const ErrorScenarios: Array<[string, IExecutionDefinition]> = [["native total", Test15]];

    it.each(Scenarios)("should correctly convert %s", (_desc, def) => {
        expect(convertTotals(def)).toMatchSnapshot();
    });
    it.each(ErrorScenarios)("should fail on converting %s", (_desc, def) => {
        expect(() => convertTotals(def)).toThrowErrorMatchingSnapshot();
    });
});
