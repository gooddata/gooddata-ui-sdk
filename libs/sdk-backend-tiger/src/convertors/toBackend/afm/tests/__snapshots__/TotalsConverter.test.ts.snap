// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertTotals > should correctly convert attributes in first dim and measure group in second dim 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert column total and row subtotal 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_opportunitysnapshot.forecastcategory_id_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "a_f_stage.isactive_id",
          "measureGroup",
        ],
      },
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_stage.isactive_id_1",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
          "a_f_opportunitysnapshot.forecastcategory_id",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "subtotal_column_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_opportunitysnapshot.forecastcategory_id_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert measure group and attribute in one dimension 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_1",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "measureGroup",
          "localAttr2",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert multiple totals with various functions 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m2_by_localAttr1_0",
    "metric": "m2",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "MAX",
    "localIdentifier": "total_max_m2_by_localAttr1_0",
    "metric": "m2",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "MIN",
    "localIdentifier": "total_min_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert no totals 1`] = `[]`;

exports[`convertTotals > should correctly convert row total and column subtotal 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_opportunitysnapshot.forecastcategory_id_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "a_f_stage.isactive_id",
          "measureGroup",
        ],
      },
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_stage.isactive_id_1",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
          "a_f_opportunitysnapshot.forecastcategory_id",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "subtotal_row_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_label.f_account.account.name_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert subtotal 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr2_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
          "measureGroup",
        ],
      },
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "localAttr1",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert subtotals with non-sorted totals 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_owner.department_id_1",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
          "a_f_opportunitysnapshot.forecastcategory_id",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_e519fa2a_86c3_4e32_8313_0c03062348j3_by_a_f_owner.department_id_1",
    "metric": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
          "a_f_opportunitysnapshot.forecastcategory_id",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert three dimensions 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
      {
        "dimensionIdentifier": "dim_2",
        "totalDimensionItems": [
          "localAttr2",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr2_2",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "localAttr1",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert three totals with the same function 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m2_by_localAttr1_0",
    "metric": "m2",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m3_by_localAttr1_0",
    "metric": "m3",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert total on dimension with measure group 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
        ],
      },
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert total on single dimension result spec 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert total without any measure group 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert two subtotals and marginal total 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_opportunitysnapshot.forecastcategory_id_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "a_f_stage.isactive_id",
          "measureGroup",
        ],
      },
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_stage.isactive_id_1",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
          "a_f_opportunitysnapshot.forecastcategory_id",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "marginal_total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_opportunitysnapshot.forecastcategory_id_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert two totals and grand total 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_label.f_account.account.name_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "a_f_owner.department_id",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_f_owner.department_id_1",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_0",
        "totalDimensionItems": [
          "a_label.f_account.account.name",
        ],
      },
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "SUM",
    "localIdentifier": "total_of_totals_sum_m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba_by_a_label.f_account.account.name_0",
    "metric": "m_973a14c4_acb1_45fb_ba52_5d96fa02f7ba",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should correctly convert two totals with different functions 1`] = `
[
  {
    "function": "SUM",
    "localIdentifier": "total_sum_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
          "measureGroup",
        ],
      },
    ],
  },
  {
    "function": "MIN",
    "localIdentifier": "total_min_m1_by_localAttr1_0",
    "metric": "m1",
    "totalDimensions": [
      {
        "dimensionIdentifier": "dim_1",
        "totalDimensionItems": [
          "localAttr3",
          "measureGroup",
        ],
      },
    ],
  },
]
`;

exports[`convertTotals > should fail on converting native total 1`] = `[Error: Tiger backend does not support native totals.]`;
